gobblin-utility/src/main/java/gobblin/util/io/GsonInterfaceAdapter.java

```diff
@@ -12,81 +12,181 @@
 import com.google.gson.JsonElement;
+import com.google.gson.JsonNull;
 import com.google.gson.JsonObject;
-import com.google.gson.JsonParseException;
-import com.google.gson.JsonSerializationContext;
-import com.google.gson.JsonSerializer;
+import com.google.gson.JsonPrimitive;
+import com.google.gson.TypeAdapter;
+import com.google.gson.TypeAdapterFactory;
+import com.google.gson.internal.Streams;
+import com.google.gson.reflect.TypeToken;
+import com.google.gson.stream.JsonReader;
+import com.google.gson.stream.JsonWriter;
 
 
 /**
- * A {@link Gson} interface adapter that makes it possible to serialize and deserialize an object
- * with an interface or abstract field with {@link Gson}.
+ * A {@link Gson} interface adapter that makes it possible to serialize and deserialize polymorphic objects.
  *
  * <p>
- *    Suppose a class <pre>MyClass</pre> contains an interface field <pre>MyInterface field</pre>. To serialize
- *    and deserialize a <pre>MyClass</pre> object, do the following:
+ *   This adapter will capture all instances of {@link #baseClass} and write them as
+ *   {"object-type":"class.name", "object-data":"data"}, allowing for correct serialization and deserialization of
+ *   polymorphic objects. The following types will not be captured by the adapter (i.e. they will be written by the
+ *   default GSON writer):
+ *   - Primitives and boxed primitives
+ *   - Arrays
+ *   - Collections
+ *   - Maps
+ *   Additionally, generic classes (e.g. class MyClass<T>) cannot be correctly decoded.
+ * </p>
  *
+ * <p>
+ *   To use:
  *    <pre>
  *          {@code
  *            MyClass object = new MyClass();
- *            Gson gson = GsonInterfaceAdapter.getGson(MyInterface.class);
+ *            Gson gson = GsonInterfaceAdapter.getGson(MyBaseClass.class);
  *            String json = gson.toJson(object);
  *            Myclass object2 = gson.fromJson(json, MyClass.class);
  *          }
  *    </pre>
  * </p>
  *
- * @author ziliu
+ * <p>
+ *   Note: a useful case is GsonInterfaceAdapter.getGson(Object.class), which will correctly serialize / deserialize
+ *   all types except for java generics.
+ * </p>
+ *
+ * @author ziliu, ibuenros
  *
  * @param <T> The interface or abstract type to be serialized and deserialized with {@link Gson}.
  */
-public class GsonInterfaceAdapter<T> implements JsonSerializer<T>, JsonDeserializer<T> {
-
-  private static final String OBJECT_TYPE = "object-type";
-  private static final String OBJECT_DATA = "object-data";
-  private static final Gson PURE_GSON = new Gson();
-
-  @Override
-  public JsonElement serialize(T object, Type interfaceType, JsonSerializationContext context) {
-    JsonObject wrapper = new JsonObject();
-    wrapper.addProperty(OBJECT_TYPE, object.getClass().getName());
-    wrapper.add(OBJECT_DATA, PURE_GSON.toJsonTree(object));
-    return wrapper;
-  }
+@RequiredArgsConstructor
+public class GsonInterfaceAdapter implements TypeAdapterFactory {
```

Comment:
In the previous version calling `GsonInterfaceAdapter.getGson(MyClass.class).toJson(new MyClass)` would create an infinite loop, so this method was useful mostly for interfaces, and not concrete classes. Also `GsonInterfaceAdapter.getGson(MyClass.class).toJson(new MyClassDerived)` would not actually write the class name, which made it very hard to correctly serialize classes with internal polymorphic fields.
The changes also add support for `Optional`.